Improved regulations build

viroulep opened the pull request:
This is happening :)

@jfly and I have been discussing about the regulations build process back when I started working the delegate reports, and two main goals emerged from this discussion:
1. The @cubing/wrc-team needed a more robust tool to easily ensure the Regulations and its translations were "well-formed", and also both repos should be more independent, so that minor revisions are live smoothly and without a PR from the WRC here.
2. The software team would like to not know anything about how the Regulations are built, and ideally would like to get rid of the two submodules (one aim behind that is to simply drop all the heavy texlive packages necessary to build the Regulations).

Hopefully this PR fully achieves both, and I managed to get rid of the full `regulations` directory.

[`wrc`](https://github.com/cubing/wca-regulations-compiler) is now able to generate pdf without any Latex, the only external dependencies being [`wkhtmltopdf`](http://wkhtmltopdf.org/downloads.html) and some fonts packages (roughly 40 mo total).
So this solves the issue of the time cost for setting up a new server, as well as the "we don't want to know about the build" issue.

Some [pages](https://github.com/cubing/worldcubeassociation.org/tree/master/regulations/pages) were "pandoc-ed" to html during the build process.
I think it's out of the scope of the Regulations compiler, and they are actually static pages that get updated only once in a while. So I decided to turn them into static pages in `app/views/regulations`.
Feel free to comment on this if you have another idea.

The [deploy](https://github.com/cubing/worldcubeassociation.org/compare/master...viroulep:improved-regulations-build?expand=1#diff-93518fda10b0403d3c5c20b4df4740a6R10) script now builds directly the default branch of the Regulations/translations repositories. These branches are guaranteed to build by the travis checks in these repositories.
(it also builds the json in view of #722)

Finally I tried to clean up the chef recipes and build process; I'm very inexperienced with this, so a careful review would be appreciated! However I destroyed and re-created the "all" VM from scratch, and everything worked like a charm (yes, I manually checked all the `/regulations` pages and pdf generated).

I think it's ready for review but there is two small last details that need to be fixed before merging this:
- [x] I'm still waiting for feedback from the Japanese delegates about the Japanese font used for the pdf, so I may have to change or add another font package to the chef regulations recipe.
- [x] I need to automatically detect if the Regulations/translations have changed from last build, so that I can run the `rebuild_regs` all the time but only actually rebuild them if they changed. I'm not sure about everything that is going on when you push a new version to `production`, what script(s) are run exactly?
I was thinking about keeping an out-of-tree file with the last versions built, what do you think about it? Any advice on where to put this file?
Any other alternatives you're thinking about?

viroulep commented on chef/site-cookbooks/wca/recipes/regulations.rb:
```diff
@@ -1,23 +1,26 @@
+
 execute "#{repo_root}/scripts/deploy.sh rebuild_regs" do
   user username
   not_if "ls #{repo_root}/WcaOnRails/app/views/regulations"
```
Is this executed every time the `production` branch is updated, or only once when the server environment is built?

jonatanklosko commented on chef/site-cookbooks/wca/recipes/regulations.rb:
```diff
@@ -1,23 +1,26 @@
+  code <<-EOH
+    wget http://download.gna.org/wkhtmltopdf/0.12/0.12.3/wkhtmltox-0.12.3_linux-generic-amd64.tar.xz -O wkhtml.tar.xz
+    tar -xf wkhtml.tar.xz -C /usr/local
+    EOH
```
Indentation =)

jonatanklosko commented:
Fixes #105 and #350.

jonatanklosko commented:
Am I right that now `RegulationsController` can be cleaned up? Also we need the appropriate routes.

jonatanklosko commented on scripts/deploy.sh:
```diff
@@ -7,17 +7,45 @@ pull_latest() {
   fi
-  mv regulations/build/regulations $regs_folder
-  rm -rf $regs_folder-todelete
+  mv regulations/build/regulations/ $regs_folder_root
```
Maybe use `$outputdir` here?

viroulep commented:
Updated with your two remarks!

I'm not sure to see how you'd like the routes to look like: do you picture a translation route with a language parameter, and static routes for scramble/history?

Also fixes #195 :)

jonatanklosko commented:
Actually I'm not sure what's the best solution as there are many pages like those with 'tnoodle', 'official' and 'files'. I'm just wondering whether all the pages loads well with the current code, have you checked it?

viroulep commented:
I just updated the deploy script to write a version file, and check it before building.

Yes I checked all the links: I checked that all the tnoodle versions were served correctly, that all the history files were there and correctly displayed, that every translations had its pdf served correctly, etc.

Actually I realized that all the controller/routing process was a bit tricky when I took a look at the regulations controller!
The solution there is quite elegant given the number of weird exceptions.

jonatanklosko commented:
Ah sorry, actually I've never looked into serving this pages process deeply. After reading HighVolgate READM.me and analyzing our code I like this =)

jonatanklosko commented on scripts/deploy.sh:
```diff
@@ -7,17 +7,88 @@ pull_latest() {
+  # This save tracked files that may have unstashed changes too
+  cp -r $regs_folder $build_folder
+
+  languages=`wrc-languages`
```
Shouldn't this be `wrc languages`?

viroulep replied:
It is [`wrc-languages`](https://github.com/cubing/wca-regulations-compiler/blob/master/setup.py#L17), although I indeed could make it an option like `wrc --languages`.

jonatanklosko replied:
Ah sorry, I see.

viroulep commented:
I just updated the PR to download the package providing the correct Japanese font.
(fonts used are [here](https://github.com/cubing/wca-regulations-compiler/blob/master/wrc/data/htmltopdf.css#L28-L38))

It should be good now, someone else should test this offline or on staging before putting this live :p

jonatanklosko commented on WcaOnRails/app/views/regulations/history/index.html.erb:
```diff
@@ -5,22 +5,23 @@
-          <li><%= link_to "2013-05-10", "./official/2013-05-10/" %> (<%= link_to "Changes", "https://github.com/cubing/wca-documents/compare/official-2013-01-01...official-2013-05-10#files_bucket" %>)</li>
-          <li><%= link_to "2013-01-01", "./official/2013-01-01/" %></li>
+          <li><%= link_to "2013-05-10", "./files/2013-05-10/" %> (<%= link_to "Changes", "https://github.com/cubing/wca-documents/compare/official-2013-01-01...official-2013-05-10#files_bucket" %>)</li>
+          <li><%= link_to "2013-01-01", "./files/2013-01-01/" %></li>
```
Why have you changed `official` to `files`? I can see the directories (*dates*) [here](https://github.com/cubing/worldcubeassociation.org/tree/regulations-data/regulations/build/regulations/history/official).

viroulep commented:
I did [here](https://github.com/viroulep/worldcubeassociation.org/tree/regulations-data).
I'll make this PR soon, actually I didn't expect this to be reviewed yet :p 

I'm still trying to create a whole new VM to properly test the change in nginx...

viroulep commented:
Updated :)
I fixed #800 as part of this one as otherwise I would have had to do the job twice :p 
Also when you'll merge this one please also merge #805 as both are required for this fix to work.

I rebased against the latest master, and appart from the latest commit the other change is that I removed [this check](https://github.com/cubing/worldcubeassociation.org/pull/747/commits/d53c2f63d6a0c932fd20a12d45c6e22d8fe7ab55#diff-668c662985415a6296057cdbc1bfc904L23), since the `rebuild_regs` script check if something needs to be rebuilt (plus this check is always true now that there is some views in the tree :s).

lgarron commented:
@viroulep Would you like me to review this? I trust your judgment, but I can review if you'd like.

viroulep commented:
Sure please do so :)

There is two main points in the PR:
* The change in the Regulations' [build](https://github.com/viroulep/worldcubeassociation.org/blob/d53c2f63d6a0c932fd20a12d45c6e22d8fe7ab55/scripts/deploy.sh#L8-L92) script
* The change in the [chef provision script](https://github.com/viroulep/worldcubeassociation.org/blob/d53c2f63d6a0c932fd20a12d45c6e22d8fe7ab55/chef/site-cookbooks/wca/recipes/regulations.rb)

The result of the first point is pretty easy/fast to check given you installed the requirements (those in the chef recipe).
The result of the second point is easy but long to check, as you would have to spin up a new server/VM from scratch...

I think @jonatanklosko reviewed the whole code part, so what this PR is waiting for is someone with time to check that this works in staging/production ;)

jfly commented on .gitmodules:
```diff
@@ -1,6 +0,0 @@
-[submodule "regulations/wca-regulations"]
```
YAYYYYYYYYYYYY

jfly commented on chef/site-cookbooks/wca/recipes/regulations.rb:
```diff
@@ -1,24 +1,27 @@
+end
+
+execute "pip install wrc --upgrade" do
+  user "root"
```
Don't bother changing this (because you'll have to retest everything), but I'm pretty sure that the `user "root"` is redundant here.

jfly commented on scripts/wca-bootstrap.sh:
```diff
@@ -63,17 +61,17 @@ EOL
   # Let the root user do the git-ing, and then chown appropriately.
   if ! [ -d $repo_root ]; then
-    git clone -b $git_branch --recursive https://github.com/cubing/worldcubeassociation.org.git $repo_root
+    git clone -b $git_branch https://github.com/cubing/worldcubeassociation.org.git $repo_root
```
Woohoo! I believe that now that we're submodule-less, we can get rid of the "initialize submodules" step in our README? https://github.com/cubing/worldcubeassociation.org#setup

viroulep replied:
done :)

jfly commented on scripts/deploy.sh:
```diff
@@ -7,17 +7,88 @@ pull_latest() {
+  translations_version=$regs_folder/translations/version
+
+  if [ -a $build_folder ]; then
+    rm -rf $build_folder
```
Is there any value in checking beforehand? You do a `rm -rf` anyways. I'd just nuke the if statement.

jfly commented on scripts/deploy.sh:
```diff
@@ -7,17 +7,88 @@ pull_latest() {
+
+  # Else we have to rebuild something
+
+  # This save tracked files that may have unstashed changes too
```
nit: "This save" -> "This saves"?

jfly commented on scripts/deploy.sh:
```diff
@@ -7,17 +7,88 @@ pull_latest() {
+
+  # Checkout data (scramble programs, history)
+  git checkout origin/regulations-data $build_folder
+  git reset HEAD $build_folder
```
My git-foo is weak. Can you explain the interaction between the last few lines? It looks like we're doing a checkout into a non-empty folder and then doing a reset... What about the untracked files we care about?

viroulep replied:
Sure, the `git checkout` part will update the "regulations/build" path to the version in `regulations-data`, without changing branch, updating the index.
Therefore you end up with a lot of file tagged for addition, since they are not present in the current branch.

The default mode for `git reset` is `--mixed`, which does not modify the content of any file, but which will reset the index (to remove the file tagged for addition).
So in this case it means "remove the file marked for addition from the index, but do not do anything to the working tree".
Therefore unstashed changes are preserved :)

jfly commented on scripts/deploy.sh:
```diff
@@ -7,17 +7,88 @@ pull_latest() {
+        inputdir=$build_folder/wca-regulations-translations/${l}
+        outputdir=$build_folder/regulations/translations/${l}
+        mkdir -p $outputdir
+        echo "Doing "${kind}" for language "${l}
```
nit: Bash supports string interpolation, so some of these quotes are unecessary:

```bash
echo "Doing ${kind} for language ${l}"
```

Also, I'd probably change "Doing" to "Generating".

viroulep replied:
Done :)

jfly commented on scripts/deploy.sh:
```diff
@@ -7,17 +7,88 @@ pull_latest() {
-    rm -rf $regs_folder-todelete
-    mv $regs_folder $regs_folder-todelete
+    # Remove previous build
+    rm -rf $regs_folder
```
The convoluted order of stuff in the old file was actually pretty intentional. We have `$outputdir` and `$regs_folder`. We want `$outputdir` to end up in place of `$regs_folder`. There are a few ways to achieve this:

## Simplest (what you did here)
1. Delete `$regs_folder`.
2. Move `$outputdir` to `$regs_folder`.

However, deleting files actually can take some time (I believe `rm -r` has to recurse through the whole directory and `unlink` a bunch of files. During this time, the regulations are in a broken state. So, I wrote things in this slightly more convoluted order:

## More convoluted
1. Move `$regs_folder` to some temp name that you plan to delete later.
2. Move `outputdir` to `$regs_folder`.
3. Now we finally delete the old `$regs_folder`.

It is my understanding that both steps 1) and 2) above should be quick, atomic operations. Then we can spend all day `unlink`-ing files in 3), but at least the website won't be down.

I may be very wrong about some of the stuff here. This also may be a classic example of premature optimization. Let me know what you think.

viroulep replied:
Interesting, I messed that point, and when restoring the convoluted version I got that:

> /vagrant @ production> rm -rf WcaOnRails/app/views/regulations-todelete
> rm: cannot remove ‘WcaOnRails/app/views/regulations-todelete/scrambles’: Directory not empty
> rm: cannot remove ‘WcaOnRails/app/views/regulations-todelete/history’: Directory not empty
> rm: cannot remove ‘WcaOnRails/app/views/regulations-todelete/translations’: Directory not empty
> /vagrant @ production> ls -al WcaOnRails/app/views/regulations-todelete/
> total 20
> drwxr-xr-x 1 vagrant vagrant 4096 Sep  6 12:11 .
> drwxr-xr-x 1 vagrant vagrant 4096 Sep  6 12:11 ..
> drwxr-xr-x 1 vagrant vagrant 4096 Sep  6 12:11 history
> drwxr-xr-x 1 vagrant vagrant 4096 Sep  6 12:11 scrambles
> drwxr-xr-x 1 vagrant vagrant 4096 Sep  6 12:11 translations
> /vagrant @ production> ls -al WcaOnRails/app/views/regulations-todelete/scrambles/
> total 8
> drwxr-xr-x 1 vagrant vagrant 4096 Sep  6 12:11 .
> drwxr-xr-x 1 vagrant vagrant 4096 Sep  6 12:11 ..

The funny part is when I list the same directory from my computer:

> [14:27] daisy:~/Documents/cube/wca/VM/worldcubeassociation.org (merged|✚7)$ lal WcaOnRails/app/views/regulations-todelete/scrambles 
> total 20
> drwxr-xr-x 3 fifi fifi 4096 sept.  6 14:04 .
> drwxr-xr-x 5 fifi fifi 4096 sept.  6 14:11 ..
> -rw-r--r-- 1 fifi fifi 4107 sept.  6 14:04 index.html.erb
> drwxr-xr-x 3 fifi fifi 4096 sept.  6 14:04 tnoodle


So `rm -rf` fails at removing an empty dir saying it's not empty... Sounds like yet another issue with the shared directories between vagrant and my computer.
I have no clue what's going on for this particular directory, I can remove other parts from the shared directories correctly...

Therefore I moved the "todelete" folder to some local directory in `/tmp`.

jfly commented on scripts/deploy.sh:
```diff
@@ -7,17 +7,88 @@ pull_latest() {
-  # cores just building the regulations.
-  regulations/make.py --num-workers 1 --wca
-  regs_folder=WcaOnRails/app/views/regulations
+  # Uses wrc, see here: https://github.com/cubing/wca-regulations-compiler
```
This is really awesome work, and I think we should merge this up immediately. That said, this doesn't exactly fit my dream of the WCA website being "unaware" of how to build the regulations. IMO, there's a good amount of logic specific to the internals of the WRC here.

How do you feel about moving some or all this logic into the wca regulations compiler at some point?

viroulep replied:
What exactly would you want to move there?

I had in mind that the website would say "i want html for this, i want pdf for that", which is the case currently, but maybe you pictured something more like "i want all the views for the regulations"?

jfly replied:
I think I was envisioning something more like "give me all the views for the regulations", maybe someday the WRC decides to generate some other format, I don't think the WCA website necessarily needs to know about that.

More important than that, though, I'm just bummed that the `rebuild_regs` function is so long. I was kind of hoping it would be be a 1 line call to `wrc`. I don't think we (the software team) need to be aware of the fact that the regulations and the translations are stored in git, or how to "Checkout data" (I would support moving `regulations-data` branch off of the WCA website and into something controlled by the WRC), or even the fact that the regulations have translations =).

That all said, I'm totally fine merging this up as is. We can discuss later what exactly we consider "WRC logic".

viroulep replied:
Ok, it will need some more discussion, lets merge this and discuss it in a separate issue/pr (#867).

viroulep commented on Vagrantfile:
```diff
@@ -10,17 +10,9 @@ Vagrant.configure(2) do |config|
-  config.vm.define "all", autostart: false do |noregs|
-    noregs.vm.provision "shell" do |s|
+  # A full development environment
+  config.vm.define "all", autostart: false do |all|
```
I didn't know :)

I'll keep it that way though, just in case we want to add another one some day.

jfly commented:
LGTM! If you decide to merge this up as is, please make an issue for us to continue the discussion in https://github.com/cubing/worldcubeassociation.org/pull/747#discussion_r77770550.
